src/arextract.py

```diff
@@ -106,6 +109,209 @@ def boxcar(slf, det, specobjs, sciframe, varframe, skyframe, crmask, scitrace):
+            msgs.work("Should give our own guess here instead of using default")
+            guess = None
+            try:
+                mask, gfit = arutils.robust_polyfit(slit_val, flux_val, fdict['deg'], function=fdict['func'], weights=weight_val, maxone=False, guesses=guess)
+            except RuntimeError:
+                msgs.warn("Bad Profile fit for object={:s}.  Skipping Optimal".format(specobjs[o].idx))
+                scitrace['opt_profile'].append(fdict)
+                continue
+            msgs.work("Consider flagging/removing CRs here")
+            # Record
+            fdict['param'] = gfit
+            fdict['mask'] = mask
+            fdict['slit_val'] = slit_val
+            fdict['flux_val'] = flux_val
+            scitrace['opt_profile'].append(fdict)
+            if msgs._debug['obj_profile']: #
+                gdp = mask == 0
+                mn = np.min(slit_val[gdp])
+                mx = np.max(slit_val[gdp])
+                xval = np.linspace(mn,mx,1000)
+                model = arutils.func_val(gfit, xval, fdict['func'])
+                import matplotlib.pyplot as plt
+                plt.clf()
+                ax = plt.gca()
+                ax.scatter(slit_val[gdp], flux_val[gdp], marker='.', s=0.7, edgecolor='none', facecolor='black')
+                ax.plot(xval, model, 'b')
+                # Gaussian too?
+                if False:
+                    fdictg = dict(func='gaussian', deg=3)
+                    maskg, gfitg = arutils.robust_polyfit(slit_val, flux_val, fdict['deg'], function=fdictg['func'], weights=weight_val, maxone=False)
+                    modelg = arutils.func_val(gfitg, xval, fdictg['func'])
+                    ax.plot(xval, modelg, 'r')
+                plt.show()
+                debugger.set_trace()
+        elif len(gdrow) > 10:  #
+            msgs.warn("Low extracted flux for obj={:s}.  Not ready for Optimal".format(specobjs[o].idx))
+            scitrace['opt_profile'].append({})
+            continue
+    # QA
+    arqa.obj_profile_qa(slf, specobjs, scitrace)
+    return scitrace['opt_profile']
+
+
+def optimal_extract(slf, det, specobjs, sciframe, varframe, skyframe, crmask, scitrace,
+                 pickle_file=None, profiles=None):
+    """ Preform optimal extraction
+    Standard Horne approach
+    Parameters
+    ----------
+    slf
+    det
+    specobjs
+    sciframe
+    varframe
+    crmask
+    scitrace
+    COUNT_LIM
+    pickle_file
+    Returns
+    -------
+    """
+    '''
+    import pickle
+    if pickle_file is not None:
+        f = open(pickle_file,'r')
+        args = pickle.load(f)
+        f.close()
+        det, specobjs, sciframe, varframe, skyframe, crmask, scitrace, tilts = args
+        slf = None
+        scitrace['opt_profile'] = profiles
+    else:
+        tilts = None
+    '''
+    # Setup
+    model_var = np.abs(skyframe + sciframe - np.sqrt(2)*slf._msrn[det-1]) + \
```

Comment:
Am pretty sure this is a poor-man's kludge for the fact that the 
low-count regime is Poission yet we are doing Gaussian stats.
I'm not sure I can rederive the sqrt(2) factor.  Might well have been
some trial and error to see what gave the closest to unbiased results
in the sky sub.